### PR TITLE
Created generic Tracker and implemented it across some components

### PR DESCRIPTION
**Pull request by munteannatan**
# Pull Request

## 🤨 Rationale

The Tracker pattern is used in multiple components with different implementations. 
This change creates a generic Tracker and replaces some of the original uses

## 👩‍💻 Implementation

Created an abstract Tracker class, extended it into an Update Tracker and a Validator, both generics, and used them in the table and table column

## 🧪 Testing

all tests pass

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/utilities/models/tracker.ts:8]: A "whim" is an unfamiliar term to me (unless it's [this definition](https://www.merriam-webster.com/dictionary/whim), which I doubt).

Comment (munteannatan) [reply packages/nimble-components/src/utilities/models/tracker.ts:8]: Yeah, I was out of ideas there and asked ChatGPT 😅, I 'm open to any suggestions

Comment (atmgrifter00) [reply packages/nimble-components/src/utilities/models/tracker.ts:8]: At the moment `trackedItemList` seems more appropriate (and for the `_whims` field it could be `trackedItemsState`?). However, see my other comment below for concern around this class in general.

Comment (atmgrifter00) [inline packages/nimble-components/src/utilities/models/tracker.ts:8]: What's the benefit of having this class in addition to the `Validator` class? The only difference in the APIs is the extra `isValid` method on the `Validator`. Do we want the `Tracker` class for cases where the boolean state represents something other than validity?

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/utilities/models/tracker.ts:1]: Providing the second type parameter (that defaults to `string` no less) seems unnecessary when the implementation relies on that type being a `boolean` (and is declared as such where it's used).

File: change/@ni-nimble-components-b275af58-4f98-4b30-a09b-f91dc2833b9b.json
Comment (rajsite) [inline change/@ni-nimble-components-b275af58-4f98-4b30-a09b-f91dc2833b9b.json:1]: @munteannatan I notice this PR is still in a draft state and does not have any questions where you @mention anyone for feedback. To me that means this PR is still in development and you are not looking for feedback yet.

If you have specific questions you need answered before moving off Draft make sure they are asked in the draft with people mentioned.

If you are ready for review then make sure to switch from Draft.

If it's still in development, that's fine, just want to make sure you are not waiting on us for anything.

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/utilities/models/tracker.ts:26]: Perhaps all of these methods should be `protected`? This would place a slight burden on the derived implementations where they would need to define their own wrapper APIs around these, but that would address some of the semantic concerns.

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (munteannatan) [reply packages/nimble-components/src/utilities/models/tracker.ts:8]: yes, the original tracker class was implemented for the update tracker in the table, we want to reuse it in the wafer map component. You can see more of the discussion in #1232. @rajsite suggested to look into reusing the same pattern for the column validator, and I tried to have the same pattern used in a generic validator class which can be also used in multiple components, not just the table columns

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (munteannatan) [reply packages/nimble-components/src/utilities/models/tracker.ts:26]: I think that will limit the usage of the tracker class on its own. Maybe using it as a private property of the validator could work?

File: change/@ni-nimble-components-b275af58-4f98-4b30-a09b-f91dc2833b9b.json
Comment (munteannatan) [reply change/@ni-nimble-components-b275af58-4f98-4b30-a09b-f91dc2833b9b.json:1]: sure thank you!

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/utilities/models/tracker.ts]: Oh sorry, I had it in my head that `Tracker` was `abstract` since the derived `UpdateTracker` was `abstract`.

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/tracker.ts:26]: Instead of returning a reference to the underlying object which could be accidentally mutated by end users can you return a shallow clone?
```suggestion
        return {...this.trackedItemsState};
```

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/tracker.ts:63]: Can we just make this private and use trackAll / untrackAll elsewhere?

File: packages/nimble-components/src/table/models/table-update-tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/table/models/table-update-tracker.ts:62]: Should be able to avoid defining this twice like the TestColumnValidator https://github.com/ni/nimble/blob/14402b0d2e2a8a56454f79aa971324bf96dea5fe/packages/nimble-components/src/table-column/base/tests/table-column.fixtures.ts#L55

File: packages/nimble-components/src/utilities/models/update-tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/update-tracker.ts:23]: I don't think these are adding significant value. Instead of developers just having to understand when queue calls happen they have to understand when track or trackAndQueue are called. Also trackAllAndQueue seems completely unused?
I think you can avoid growing the api surface in these cases

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (rajsite) [reply packages/nimble-components/src/utilities/models/tracker.ts:26]: When done this should also be changed to a method instead of an accessor, i.e. getTrackedItems()

**Review by jattasNI — COMMENTED** [review]
I'm going to defer to @rajsite on this PR since he's been more involved in the brainstorming. No need to wait for my review to submit (though feel free to pull me in if there are questions for me). Thanks for the contribution!

File: packages/nimble-components/src/utilities/models/update-tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/update-tracker.ts:11]: Constructor isn't doing anything anymore, think you can leave it off and calls will go right to the parent

File: packages/nimble-components/src/utilities/models/validator.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/validator.ts:12]: Same as the other, constructor isn't doing anything anymore, think you can leave it off and calls will go right to the parent

File: packages/nimble-components/src/utilities/models/tracker.ts
Comment (rajsite) [inline packages/nimble-components/src/utilities/models/tracker.ts:33]: Reading how it looks in source, can we rename this to `isTracked`? I think it matches the pattern of all/any/none-Tracked better for checking the tracked item status.

**Review by rajsite — APPROVED** [review]

**Review by atmgrifter00 — APPROVED** [review]

**Pull request merged**
